File: src/main/java/de/pflugradts/pwman3/domain/model/ddd/AggregateRoot.java
package de.pflugradts.pwman3.domain.model.ddd;

import java.util.List;

/**
 * Denotes a Domain Entity that serves as an AggregateRoot in Domain-driven Design.
 */
public interface AggregateRoot {
    List<DomainEvent> getDomainEvents();
    default void registerDomainEvent(final DomainEvent domainEvent) {
        getDomainEvents().add(domainEvent);
    }
    default void clearDomainEvents() {
        getDomainEvents().clear();
    }
}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/ddd/AggregateRoot.kt
package de.pflugradts.passbird.domain.model.ddd

import java.util.Collections

abstract class AggregateRoot {
    private val domainEvents = mutableListOf<DomainEvent>()
    fun registerDomainEvent(domainEvent: DomainEvent) {
        domainEvents.add(domainEvent)
    }
    fun clearDomainEvents() {
        domainEvents.clear()
    }
    fun getDomainEvents(): List<DomainEvent> = Collections.unmodifiableList(domainEvents)
}
